### PR TITLE
Write operator actions to audit log

### PR DESCRIPTION
**Pull request by mpolden**
All mutable requests to the following paths will be written to the audit log:

* `/controller/v1/`
* `/os/v1/`
* `/zone/v2/`

Those paths cover the most important operator actions, such as:

* (De)activation of maintenance jobs on controller
* Version confidence overrides
* Scheduling OS or firmware upgrades
* Proxied calls to config servers: Feature flag changes, (de)activation of
  node repository maintenance jobs, node state changes, scheduling of node reboots etc.

We can also consider adding node (de)provisioning (`/provision/v2/`), but that
needs to be changed in internal code.

Future handlers that require audit logging can simply extend
`AuditLoggingRequestHandler`.

FYI @hakonhall

Will update operations doc after this is merged.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/auditlog/AuditLoggingRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    private final AuditLogger auditLogger;
+
+    public AuditLoggingRequestHandler(Context ctx, AuditLogger auditLogger) {
+        super(ctx);
+        this.auditLogger = auditLogger;
+    }
+
+    @Override
+    public HttpResponse handle(HttpRequest request) {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/auditlog/AuditLoggingRequestHandler.java]: Make final

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java [hunk 2]
```diff
@@ -429,6 +436,17 @@ public LongStream getLogChunkIds(ApplicationId id, JobType type) {
...
     }
 
+    // -------------- Audit log -----------------------------------------------
+
+    public AuditLog readAuditLog() {
+        return readSlime(auditLogPath()).map(auditLogSerializer::fromSlime)
+                                        .orElse(AuditLog.empty);
+    }
+
+    public void writeAuditLog(AuditLog log) {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java]: Should this take the lock as argument such that nobody accidentally forgets to lock? Or at least javadoc it ...

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/auditlog/AuditLogger.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    private final CuratorDb db;
+    private final Clock clock;
+
+    public AuditLogger(CuratorDb db, Clock clock) {
+        this.db = Objects.requireNonNull(db, "db must be non-null");
+        this.clock = Objects.requireNonNull(clock, "clock must be non-null");
+    }
+
+    /** Read the current audit log */
+    public AuditLog get() {
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/auditlog/AuditLogger.java]: 1. I think get() is odd when it returns something else than itself. 
2. I suggest calling this something with read so it is clear that it is not cheap.

**Review by bratseth — APPROVED** [review]
Great! Consider my comments.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java (hunk 2 above)
Comment (mpolden) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/CuratorDb.java]: Considered, but `AuditLogger` should really remain the only writer...

And we don't do it for the similar cases in this class, instead we try to enforce such requirements one level up. E.g. `ApplicationController#store(LockedApplication)` or `RotationRepository#getOrAssignRotation(Application, RotationLock)`.

**Comment by mpolden** [comment]
PTAL

**Review by bratseth — APPROVED** [review]

**Pull request merged**